File: keyboard/src/main/java/com/carbit/inappkeyboard/keyboard/InAppKeyboardView.java
package com.carbit.inappkeyboard.keyboard;

import android.content.Context;
import android.content.res.Configuration;
import android.os.Handler;
import android.os.Looper;
import android.text.Editable;
import android.text.InputType;
import android.util.AttributeSet;
import android.util.Log;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.graphics.Color;

import androidx.annotation.VisibleForTesting;

import java.util.ArrayList;
import java.util.List;

/**
 * A minimal in-app (non-IME) soft keyboard.
 *
 * Key idea: we do NOT rely on Android's InputMethodService, so this works inside a Presentation
 * shown on a VirtualDisplay (where the system IME often won't appear).
 */
public class InAppKeyboardView extends LinearLayout {
    private static final String TAG = "InAppKeyboardView";
    private static final boolean DEBUG = true;

    public enum Layout { EN, ZH_PINYIN, FR, AR, SYMBOLS, NUMERIC }

    public enum InputMode { AUTO, TEXT, NUMBER, PASSWORD }

    public interface OnLayoutChangedListener {
        void onLayoutChanged(Layout layout);
    }

    public interface OnCommitTextListener {
        boolean onCommitText(Layout layout, String text);
    }

    public interface OnBackspaceListener {
        boolean onBackspace(Layout layout);
    }

    public interface OnSpaceListener {
        boolean onSpace(Layout layout);
    }

    private boolean inputModeLocked = false;
    private InputMode inputMode = InputMode.AUTO;

    private Editable target;
    private EditText boundEditText;
    private ICandidateBar candidateBar;
    private PinyinDecoder pinyinDecoder;
    private PinyinImeSession pinyinSession;

    private boolean isLandscape = false;
    private boolean isUltraWide = false;

    private Layout currentLayout = Layout.EN;

    private OnLayoutChangedListener onLayoutChangedListener;
    private OnCommitTextListener onCommitTextListener;
    private OnBackspaceListener onBackspaceListener;
    private OnSpaceListener onSpaceListener;

    private boolean shift = false;
    private final Handler handler = new Handler(Looper.getMainLooper());

    public InAppKeyboardView(Context context) {
        this(context, null);
    }

    public InAppKeyboardView(Context context, AttributeSet attrs) {
        super(context, attrs);
        setOrientation(VERTICAL);
        setGravity(Gravity.CENTER_HORIZONTAL);
        updateSizeFlags(getWidth(), getHeight());
        rebuild();
    }

    public InputMode getInputMode() {
        return inputMode;
    }

    public void setInputMode(InputMode value) {
        inputMode = value;
        inputModeLocked = (value != InputMode.AUTO);
        applyInputModeIfNeeded();
    }

    private void setInferredInputMode(InputMode value) {
        inputMode = value;
        inputModeLocked = false;
        applyInputModeIfNeeded();
    }

    public Editable getTarget() {
        return target;
    }

    public void setTarget(Editable target) {
        this.target = target;
    }

    public Layout getCurrentLayout() {
        return currentLayout;
    }

    public OnLayoutChangedListener getOnLayoutChangedListener() {
        return onLayoutChangedListener;
    }

    public void setOnLayoutChangedListener(OnLayoutChangedListener listener) {
        this.onLayoutChangedListener = listener;
    }

    public OnCommitTextListener getOnCommitTextListener() {
        return onCommitTextListener;
    }

    public void setOnCommitTextListener(OnCommitTextListener listener) {
        this.onCommitTextListener = listener;
    }

    public OnBackspaceListener getOnBackspaceListener() {
        return onBackspaceListener;
    }

    public void setOnBackspaceListener(OnBackspaceListener listener) {
        this.onBackspaceListener = listener;
    }

    public OnSpaceListener getOnSpaceListener() {
        return onSpaceListener;
    }

    public void setOnSpaceListener(OnSpaceListener listener) {
        this.onSpaceListener = listener;
    }

    public void attachTarget(Editable editable) {
        this.target = editable;
    }

    public void attachTo(EditText editText) {
        attachTo(editText, null);
    }

    public void attachTo(EditText editText, ICandidateBar candidateBar) {
        this.boundEditText = editText;
        this.candidateBar = candidateBar;
        attachTarget(editText.getText());

        try {
            editText.setShowSoftInputOnFocus(false);
        } catch (Throwable ignored) {
        }

        if (!inputModeLocked) {
            setInferredInputMode(inferInputMode(editText.getInputType()));
        }

        wireBuiltInPinyinIfNeeded();
        applyInputModeIfNeeded();
    }

    /** Ensure keys are built (e.g., after attachment/visibility changes). */
    public void ensureBuilt() {
        if (getChildCount() == 0) {
            rebuild();
        }
    }

    public void release() {
        try {
            if (pinyinDecoder != null) pinyinDecoder.close();
        } catch (Throwable ignored) {
        }
        pinyinDecoder = null;
        pinyinSession = null;
    }

    public void setLayout(Layout layout) {
        if (currentLayout == layout) return;
        currentLayout = layout;
        shift = false;
        rebuild();
        if (onLayoutChangedListener != null) onLayoutChangedListener.onLayoutChanged(currentLayout);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        boolean beforeLandscape = isLandscape;
        boolean beforeUltraWide = isUltraWide;
        updateSizeFlags(w, h);
        if (beforeLandscape != isLandscape || beforeUltraWide != isUltraWide) {
            rebuild();
        }
    }

    private void updateSizeFlags(int w, int h) {
        Configuration cfg = getResources().getConfiguration();
        isLandscape = cfg.orientation == Configuration.ORIENTATION_LANDSCAPE || (w > 0 && h > 0 && w > h);
        float aspect = (w > 0 && h > 0) ? (float) w / (float) h : 0f;
        isUltraWide = aspect >= 2.0f || w >= dp(1000);
    }

    private InputMode inferInputMode(int inputType) {
        int variation = inputType & InputType.TYPE_MASK_VARIATION;
        int klass = inputType & InputType.TYPE_MASK_CLASS;

        boolean isPassword =
                variation == InputType.TYPE_TEXT_VARIATION_PASSWORD
                        || variation == InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD
                        || variation == InputType.TYPE_TEXT_VARIATION_WEB_PASSWORD
                        || variation == InputType.TYPE_NUMBER_VARIATION_PASSWORD;

        if (isPassword) return InputMode.PASSWORD;
        if (klass == InputType.TYPE_CLASS_NUMBER || klass == InputType.TYPE_CLASS_PHONE) {
            return InputMode.NUMBER;
        }
        return InputMode.TEXT;
    }

    private void applyInputModeIfNeeded() {
        switch (inputMode) {
            case AUTO:
                break;
            case NUMBER:
                if (currentLayout != Layout.NUMERIC) setLayout(Layout.NUMERIC);
                break;
            case PASSWORD:
                if (currentLayout != Layout.EN) setLayout(Layout.EN);
                if (candidateBar != null) candidateBar.clear();
                break;
            case TEXT:
                if (currentLayout == Layout.NUMERIC || currentLayout == Layout.SYMBOLS) {
                    setLayout(Layout.EN);
                }
                break;
        }
    }

    private void wireBuiltInPinyinIfNeeded() {
        final EditText et = boundEditText;
        final ICandidateBar bar = candidateBar;

        if (bar == null || inputMode != InputMode.TEXT) {
            pinyinSession = null;
            try {
                if (pinyinDecoder != null) pinyinDecoder.close();
            } catch (Throwable ignored) {
            }
            pinyinDecoder = null;
            return;
        }

        if (pinyinDecoder == null) pinyinDecoder = new PinyinDecoder(getContext());
        if (pinyinSession == null) pinyinSession = new PinyinImeSession(pinyinDecoder);

        final ITextCommitTarget commitTarget = new ITextCommitTarget() {
            @Override
            public void insert(String text) {
                int pos = Math.max(et.getSelectionStart(), et.getText().length());
                et.getText().insert(pos, text);
            }
        };

        final Runnable refreshCandidates = new Runnable() {
            @Override
            public void run() {
                if (pinyinSession == null) return;
                if (!pinyinSession.hasComposing()) {
                    bar.clear();
                    return;
                }
                pinyinSession.bindCandidateClicks(commitTarget, bar);
            }
        };

        onCommitTextListener = new OnCommitTextListener() {
            @Override
            public boolean onCommitText(Layout layout, String text) {
                if (layout == Layout.ZH_PINYIN && pinyinSession != null) {
                    pinyinSession.onCommitChar(text, bar);
                    refreshCandidates.run();
                    return true;
                }
                return false;
            }
        };

        onBackspaceListener = new OnBackspaceListener() {
            @Override
            public boolean onBackspace(Layout layout) {
                if (layout == Layout.ZH_PINYIN && pinyinSession != null) {
                    boolean consumed = pinyinSession.onBackspace(bar);
                    if (consumed) refreshCandidates.run();
                    return consumed;
                }
                return false;
            }
        };

        onSpaceListener = new OnSpaceListener() {
            @Override
            public boolean onSpace(Layout layout) {
                if (layout == Layout.ZH_PINYIN && pinyinSession != null) {
                    boolean consumed = pinyinSession.onSpaceCommitBest(commitTarget, bar);
                    if (consumed) refreshCandidates.run();
                    return consumed;
                }
                return false;
            }
        };

        final OnLayoutChangedListener existing = onLayoutChangedListener;
        onLayoutChangedListener = new OnLayoutChangedListener() {
            @Override
            public void onLayoutChanged(Layout layout) {
                if (existing != null) existing.onLayoutChanged(layout);
                if (layout != Layout.ZH_PINYIN && pinyinSession != null) {
                    pinyinSession.clear();
                    bar.clear();
                }
            }
        };
    }

    private void rebuild() {
        removeAllViews();
        if (DEBUG) Log.d(TAG, "rebuild layout=" + currentLayout + " view=" + System.identityHashCode(this));
        switch (currentLayout) {
            case EN:
                buildEnQwerty();
                break;
            case ZH_PINYIN:
                buildZhPinyin();
                break;
            case FR:
                buildFrAzerty();
                break;
            case AR:
                buildArabic();
                break;
            case SYMBOLS:
                buildSymbols();
                break;
            case NUMERIC:
                buildNumeric();
                break;
        }
    }

    private void buildEnQwerty() {
        buildFromAskXml("ask_layouts/en_qwerty.xml");
    }

    private void buildZhPinyin() {
        buildFromAskXml("ask_layouts/en_qwerty.xml");
    }

    private void buildFrAzerty() {
        buildFromAskXml("ask_layouts/fr_azerty.xml");
    }

    private void buildArabic() {
        buildFromAskXml("ask_layouts/ar_qwerty.xml");
    }

    private void buildSymbols() {
        addRow(list("1", "2", "3", "4", "5", "6", "7", "8", "9", "0"));
        addRow(list("@", "#", "$", "%", "&", "*", "-", "+", "(", ")"));
        addRow(list("abc", "_", "\"", "'", ":", ";", "!", "?", "⌫"));
        addRow(list("lang", "space", "enter"));
    }

    private void buildNumeric() {
        addRow(list("1", "2", "3"));
        addRow(list("4", "5", "6"));
        addRow(list("7", "8", "9"));
        addRow(list("123", "0", "⌫"));
        addRow(list("lang", "enter"));
    }

    private static List<String> list(String... items) {
        List<String> l = new ArrayList<>();
        for (String s : items) l.add(s);
        return l;
    }

    private void buildFromAskXml(String assetPath) {
        try {
            AskXmlKeyboardParser.Layout layout = AskXmlKeyboardParser.parseAsset(getContext().getAssets(), assetPath);

            int totalKeys = 0;
            for (List<AskXmlKeyboardParser.Key> row : layout.rows) {
                List<String> labels = new ArrayList<>();
                for (AskXmlKeyboardParser.Key key : row) {
                    String label = null;
                    if (key.code != null) {
                        if (key.code == -1) label = "⇧";
                        else if (key.code == -5) label = "⌫";
                        else label = key.label != null ? key.label : String.valueOf((char) (int) key.code);
                    } else if (key.label != null && !key.label.isEmpty()) {
                        label = key.label;
                    }
                    if (label != null) labels.add(label);
                }
                if (!labels.isEmpty()) {
                    totalKeys += labels.size();
                    addRow(labels);
                    if (DEBUG) Log.d(TAG, "addRow labels=" + labels.size() + " totalKeys=" + totalKeys);
                }
            }

            if (totalKeys == 0) {
                // Fallback: minimal qwerty when ASK parsing yields no keys.
                addRow(list("q","w","e","r","t","y","u","i","o","p"));
                addRow(list("a","s","d","f","g","h","j","k","l"));
                addRow(list("⇧","z","x","c","v","b","n","m","⌫"));
            }

            addRow(list("lang", "123", "space", "enter"));
            if (DEBUG) Log.d(TAG, "rebuild done childCount=" + getChildCount() + " view=" + System.identityHashCode(this));
            requestLayout();
            invalidate();
        } catch (Exception e) {
            throw new RuntimeException("Failed to parse keyboard layout: " + assetPath, e);
        }
    }

    private void addRow(List<String> keys) {
        LinearLayout row = new LinearLayout(getContext());
        row.setOrientation(HORIZONTAL);
        row.setGravity(Gravity.CENTER);
        row.setLayoutParams(new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT));

        int keyHeight = isLandscape ? dp(36) : dp(44);
        int fixedKeyWidth = isUltraWide ? dp(56) : 0;

        for (String label : keys) {
            Button btn = new Button(getContext());
            btn.setText(resolveKeyLabel(label));
            btn.setAllCaps(false);
            btn.setMinHeight(keyHeight);
            btn.setMinimumHeight(keyHeight);
            // Make keys visible against dark background
            btn.setBackgroundColor(Color.parseColor("#8E7CC3"));
            btn.setTextColor(Color.WHITE);

            if (isUltraWide) {
                int w = "space".equals(label) ? fixedKeyWidth * 4 : "enter".equals(label) || "lang".equals(label) ? fixedKeyWidth * 2 : fixedKeyWidth;
                LayoutParams lp = new LayoutParams(w, LayoutParams.WRAP_CONTENT);
                lp.setMarginStart(dp(2));
                lp.setMarginEnd(dp(2));
                btn.setLayoutParams(lp);
            } else {
                float weight = "space".equals(label) ? 3f : 1f;
                LayoutParams lp = new LayoutParams(0, LayoutParams.WRAP_CONTENT, weight);
                lp.setMarginStart(dp(2));
                lp.setMarginEnd(dp(2));
                btn.setLayoutParams(lp);
            }

            if ("⌫".equals(label)) {
                setupBackspaceRepeater(btn);
            } else {
                btn.setOnClickListener(new OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        onKey(label);
                    }
                });
            }

            row.addView(btn);
        }

        addView(row);
    }

    private String resolveKeyLabel(String label) {
        if ("space".equals(label)) return "Space";
        if ("enter".equals(label)) return "Enter";
        if ("abc".equals(label)) return "ABC";
        if ("123".equals(label)) return (currentLayout == Layout.NUMERIC) ? "#+=" : "123";
        if ("lang".equals(label)) {
            switch (currentLayout) {
                case EN: return "EN";
                case ZH_PINYIN: return "中";
                case FR: return "FR";
                case AR: return "AR";
                case SYMBOLS: return "#";
                case NUMERIC: return "123";
            }
        }
        return label;
    }

    @VisibleForTesting
    public void injectKey(String label) {
        onKey(label);
    }

    private void onKey(String label) {
        if (target == null) return;

        if ("⌫".equals(label)) {
            boolean consumed = onBackspaceListener != null && onBackspaceListener.onBackspace(currentLayout);
            if (!consumed) {
                int len = target.length();
                if (len > 0) target.delete(len - 1, len);
            }
        } else if ("enter".equals(label)) {
            target.append("\n");
        } else if ("space".equals(label)) {
            boolean consumed = onSpaceListener != null && onSpaceListener.onSpace(currentLayout);
            if (!consumed) target.append(" ");
        } else if ("⇧".equals(label)) {
            shift = !shift;
        } else if ("123".equals(label)) {
            setLayout(Layout.SYMBOLS);
        } else if ("abc".equals(label)) {
            if (inputMode == InputMode.NUMBER) setLayout(Layout.NUMERIC);
            else setLayout(Layout.EN);
        } else if ("lang".equals(label)) {
            if (inputMode == InputMode.NUMBER) {
                setLayout(Layout.NUMERIC);
                return;
            }
            if (inputMode == InputMode.PASSWORD) {
                setLayout(Layout.EN);
                return;
            }
            Layout next;
            switch (currentLayout) {
                case EN: next = Layout.ZH_PINYIN; break;
                case ZH_PINYIN: next = Layout.FR; break;
                case FR: next = Layout.AR; break;
                case AR:
                case SYMBOLS:
                case NUMERIC: next = Layout.EN; break;
                default: next = Layout.EN; break;
            }
            setLayout(next);
        } else {
            String ch = shift ? label.toUpperCase() : label;
            boolean consumed = onCommitTextListener != null && onCommitTextListener.onCommitText(currentLayout, ch);
            if (!consumed) target.append(ch);
            if (shift) shift = false;
        }
    }

    private void setupBackspaceRepeater(final Button btn) {
        final long initialDelayMs = 250L;
        final long repeatDelayMs = 50L;

        final boolean[] repeating = { false };
        final Runnable repeatRunnable = new Runnable() {
            @Override
            public void run() {
                if (!repeating[0]) return;
                onKey("⌫");
                handler.postDelayed(this, repeatDelayMs);
            }
        };

        btn.setOnTouchListener(new OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                switch (event.getActionMasked()) {
                    case MotionEvent.ACTION_DOWN:
                        onKey("⌫");
                        repeating[0] = true;
                        handler.postDelayed(repeatRunnable, initialDelayMs);
                        return true;
                    case MotionEvent.ACTION_UP:
                    case MotionEvent.ACTION_CANCEL:
                        repeating[0] = false;
                        handler.removeCallbacks(repeatRunnable);
                        return true;
                    default:
                        return false;
                }
            }
        });
    }

    private int dp(int v) {
        return (int) (v * getResources().getDisplayMetrics().density);
    }
}

File: keyboard/src/main/java/com/carbit/inappkeyboard/keyboard/InAppKeyboardView.kt
package com.carbit.inappkeyboard.keyboard

import android.content.Context
import android.content.res.Configuration
import android.text.Editable
import android.text.InputType
import android.util.AttributeSet
import android.os.Handler
import android.os.Looper
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.LinearLayout

/**
 * A minimal in-app (non-IME) soft keyboard.
 *
 * Key idea: we do NOT rely on Android's InputMethodService, so this works inside a Presentation
 * shown on a VirtualDisplay (where the system IME often won't appear).
 */
class InAppKeyboardView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
) : LinearLayout(context, attrs) {

    /**
     * Supported layouts (no candidate/suggestion system).
     *
     * - EN: QWERTY
     * - ZH: Pinyin mode (same letters as EN, different label + can add CN punctuation later)
     * - FR: AZERTY
     * - AR: Arabic basic
     */
    enum class Layout { EN, ZH_PINYIN, FR, AR, SYMBOLS, NUMERIC }

    enum class InputMode { AUTO, TEXT, NUMBER, PASSWORD }

    private var inputModeLocked: Boolean = false
    private var _inputMode: InputMode = InputMode.AUTO

    /**
     * Input mode policy.
     *
     * - Default is [AUTO].
     * - If you set this to a non-AUTO value, it becomes "locked" and [attachTo] will not override it.
     * - If you set it back to AUTO, [attachTo] will infer mode from EditText.inputType.
     */
    var inputMode: InputMode
        get() = _inputMode
        set(value) {
            _inputMode = value
            inputModeLocked = value != InputMode.AUTO
            applyInputModeIfNeeded()
        }

    private fun setInferredInputMode(value: InputMode) {
        // Internal helper: inferred from EditText.inputType; should not lock.
        _inputMode = value
        inputModeLocked = false
        applyInputModeIfNeeded()
    }

    var target: Editable? = null

    private var boundEditText: EditText? = null
    private var candidateBar: ICandidateBar? = null

    // Optional built-in pinyin wiring (only used when attachTo(..., candidateBar=...)).
    private var pinyinDecoder: PinyinDecoder? = null
    private var pinyinSession: PinyinImeSession? = null

    private var isLandscape = false
    private var isUltraWide = false

    var currentLayout: Layout = Layout.EN
        private set

    /** Notifies host to adjust target view direction (e.g. RTL for Arabic). */
    var onLayoutChanged: ((Layout) -> Unit)? = null

    /**
     * Optional interception hooks.
     * Return true to consume and prevent default Editable writes.
     */
    var onCommitText: ((layout: Layout, text: String) -> Boolean)? = null
    var onBackspace: ((layout: Layout) -> Boolean)? = null
    var onSpace: ((layout: Layout) -> Boolean)? = null

    init {
        orientation = VERTICAL
        gravity = Gravity.CENTER_HORIZONTAL
        updateSizeFlags(width, height)
        rebuild()
    }

    /**
     * Lowest-level binding: just provide an [Editable] to receive committed characters.
     *
     * If you want auto input-type handling + optional pinyin candidate wiring, use [attachTo].
     */
    fun attachTarget(editable: Editable?) {
        target = editable
    }

    /**
     * High-level binding intended for library consumers.
     *
     * - Disables the system IME (showSoftInputOnFocus=false)
     * - Attaches target Editable
     * - Auto-selects a keyboard layout for NUMBER/PASSWORD/TEXT
     * - Optionally wires Pinyin composing + candidate commit when [candidateBar] is provided
     */
    fun attachTo(editText: EditText, candidateBar: ICandidateBar? = null) {
        boundEditText = editText
        this.candidateBar = candidateBar
        attachTarget(editText.text)

        // Prevent system IME.
        try {
            editText.showSoftInputOnFocus = false
        } catch (_: Throwable) {
            // ignore on older API
        }

        // Infer input mode from inputType unless consumer explicitly locked inputMode.
        if (!inputModeLocked) {
            setInferredInputMode(inferInputMode(editText.inputType))
        }

        wireBuiltInPinyinIfNeeded()
        applyInputModeIfNeeded()
    }

    /** Release JNI decoder resources if [attachTo] created them. */
    fun release() {
        try {
            pinyinDecoder?.close()
        } catch (_: Throwable) {
        }
        pinyinDecoder = null
        pinyinSession = null
    }

    fun setLayout(layout: Layout) {
        if (currentLayout == layout) return
        currentLayout = layout
        shift = false
        rebuild()
        onLayoutChanged?.invoke(currentLayout)
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        val beforeLandscape = isLandscape
        val beforeUltraWide = isUltraWide
        updateSizeFlags(w, h)
        if (beforeLandscape != isLandscape || beforeUltraWide != isUltraWide) {
            rebuild()
        }
    }

    private fun updateSizeFlags(w: Int, h: Int) {
        val cfg = resources.configuration
        isLandscape = cfg.orientation == Configuration.ORIENTATION_LANDSCAPE || (w > 0 && h > 0 && w > h)
        // "Ultra-wide" heuristic: very wide aspect ratio OR large width.
        val aspect = if (w > 0 && h > 0) w.toFloat() / h.toFloat() else 0f
        isUltraWide = aspect >= 2.0f || w >= dp(1000)
    }

    private fun inferInputMode(inputType: Int): InputMode {
        val variation = inputType and InputType.TYPE_MASK_VARIATION
        val klass = inputType and InputType.TYPE_MASK_CLASS

        val isPassword =
            variation == InputType.TYPE_TEXT_VARIATION_PASSWORD ||
                variation == InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD ||
                variation == InputType.TYPE_TEXT_VARIATION_WEB_PASSWORD ||
                variation == InputType.TYPE_NUMBER_VARIATION_PASSWORD

        if (isPassword) return InputMode.PASSWORD

        if (klass == InputType.TYPE_CLASS_NUMBER || klass == InputType.TYPE_CLASS_PHONE) {
            return InputMode.NUMBER
        }

        return InputMode.TEXT
    }

    private fun applyInputModeIfNeeded() {
        // Called when inputMode changes or when attachTo binds an EditText.
        when (inputMode) {
            InputMode.AUTO -> Unit // will be resolved in attachTo
            InputMode.NUMBER -> {
                // prefer numeric keypad
                if (currentLayout != Layout.NUMERIC) setLayout(Layout.NUMERIC)
            }
            InputMode.PASSWORD -> {
                // Avoid pinyin/candidates by default.
                // Always start from EN for password fields.
                if (currentLayout != Layout.EN) setLayout(Layout.EN)
                candidateBar?.clear()
            }
            InputMode.TEXT -> {
                // For normal text fields, avoid staying in numeric/symbol-only layouts.
                if (currentLayout == Layout.NUMERIC || currentLayout == Layout.SYMBOLS) {
                    setLayout(Layout.EN)
                }
            }
        }
    }

    private fun wireBuiltInPinyinIfNeeded() {
        val et = boundEditText ?: return
        val bar = candidateBar

        // Only wire pinyin when we have a candidate bar and the field isn't password/number.
        if (bar == null || inputMode != InputMode.TEXT) {
            pinyinSession = null
            try { pinyinDecoder?.close() } catch (_: Throwable) {}
            pinyinDecoder = null
            return
        }

        if (pinyinDecoder == null) {
            pinyinDecoder = PinyinDecoder(context)
        }
        if (pinyinSession == null) {
            pinyinSession = PinyinImeSession(pinyinDecoder!!)
        }

        val commitTarget = object : ITextCommitTarget {
            override fun insert(text: String) {
                et.text.insert(et.selectionStart.coerceAtLeast(et.text.length), text)
            }
        }

        fun refreshCandidates() {
            val session = pinyinSession ?: return
            if (!session.hasComposing()) {
                bar.clear()
                return
            }
            session.bindCandidateClicks(commitTarget, bar)
        }

        // NOTE: these callbacks remain overridable by host. We only provide sensible defaults.
        onCommitText = { layout, text ->
            val session = pinyinSession
            if (layout == Layout.ZH_PINYIN && session != null) {
                session.onCommitChar(text, bar)
                refreshCandidates()
                true
            } else {
                false
            }
        }

        onBackspace = { layout ->
            val session = pinyinSession
            if (layout == Layout.ZH_PINYIN && session != null) {
                val consumed = session.onBackspace(bar)
                if (consumed) refreshCandidates()
                consumed
            } else {
                false
            }
        }

        onSpace = { layout ->
            val session = pinyinSession
            if (layout == Layout.ZH_PINYIN && session != null) {
                val consumed = session.onSpaceCommitBest(commitTarget, bar)
                if (consumed) refreshCandidates()
                consumed
            } else {
                false
            }
        }

        // Clear pinyin state when leaving ZH mode.
        val existing = onLayoutChanged
        onLayoutChanged = { layout ->
            existing?.invoke(layout)
            val session = pinyinSession
            if (layout != Layout.ZH_PINYIN && session != null) {
                session.clear()
                bar.clear()
            }
        }
    }

    private fun rebuild() {
        removeAllViews()
        when (currentLayout) {
            Layout.EN -> buildEnQwerty()
            Layout.ZH_PINYIN -> buildZhPinyin()
            Layout.FR -> buildFrAzerty()
            Layout.AR -> buildArabic()
            Layout.SYMBOLS -> buildSymbols()
            Layout.NUMERIC -> buildNumeric()
        }
    }

    private fun buildEnQwerty() {
        buildFromAskXml("ask_layouts/en_qwerty.xml")
    }

    private fun buildZhPinyin() {
        // Pinyin uses latin letters. We reuse the English layout.
        buildFromAskXml("ask_layouts/en_qwerty.xml")
    }

    private fun buildFrAzerty() {
        buildFromAskXml("ask_layouts/fr_azerty.xml")
    }

    private fun buildArabic() {
        buildFromAskXml("ask_layouts/ar_qwerty.xml")
    }

    private fun buildSymbols() {
        addRow(listOf("1","2","3","4","5","6","7","8","9","0"))
        addRow(listOf("@","#","$","%","&","*","-","+","(",")"))
        addRow(listOf("abc","_","\"","'",":",";","!","?","⌫"))
        addRow(listOf("lang","space","enter"))
    }

    private fun buildNumeric() {
        // Simple numeric keypad-style layout.
        addRow(listOf("1","2","3"))
        addRow(listOf("4","5","6"))
        addRow(listOf("7","8","9"))
        addRow(listOf("123","0","⌫"))
        addRow(listOf("lang","enter"))
    }

    private fun buildFromAskXml(assetPath: String) {
        // Parse ASK xml keyboard and render keys.
        val layout = AskXmlKeyboardParser.parseAsset(context.assets, assetPath)

        layout.rows.forEach { row ->
            // Convert ASK key definitions into our labels.
            val labels = row.mapNotNull { key ->
                when (key.code) {
                    -1 -> "⇧"
                    -5 -> "⌫"
                    null -> null
                    else -> {
                        // Prefer explicit label when provided.
                        key.label ?: key.code.toChar().toString()
                    }
                }
            }
            if (labels.isNotEmpty()) addRow(labels)
        }

        // Add our bottom utility row.
        addRow(listOf("lang","123","space","enter"))
    }

    private fun addRow(keys: List<String>) {
        val row = LinearLayout(context).apply {
            orientation = HORIZONTAL
            gravity = Gravity.CENTER
            layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT)
        }

        val keyHeight = if (isLandscape) dp(36) else dp(44)
        val fixedKeyWidth = if (isUltraWide) dp(56) else 0

        keys.forEach { label ->
            val btn = Button(context).apply {
                text = when (label) {
                    "space" -> "Space"
                    "enter" -> "Enter"
                    "abc" -> "ABC"
                    "123" -> if (currentLayout == Layout.NUMERIC) "#+=" else "123"
                    "lang" -> when (currentLayout) {
                        Layout.EN -> "EN"
                        Layout.ZH_PINYIN -> "中"
                        Layout.FR -> "FR"
                        Layout.AR -> "AR"
                        Layout.SYMBOLS -> "#"
                        Layout.NUMERIC -> "123"
                    }
                    else -> label
                }
                isAllCaps = false
                minHeight = keyHeight
                minimumHeight = keyHeight
            }

            btn.layoutParams = if (isUltraWide) {
                // Fixed widths so keys don't become comically wide on ultra-wide screens.
                val w = when (label) {
                    "space" -> fixedKeyWidth * 4
                    "enter" -> fixedKeyWidth * 2
                    "lang" -> fixedKeyWidth * 2
                    else -> fixedKeyWidth
                }
                LayoutParams(w, LayoutParams.WRAP_CONTENT).apply {
                    marginStart = dp(2)
                    marginEnd = dp(2)
                }
            } else {
                // Normal weight-based stretch for phones/tablets.
                val weight = when (label) {
                    "space" -> 3f
                    else -> 1f
                }
                LayoutParams(0, LayoutParams.WRAP_CONTENT, weight).apply {
                    marginStart = dp(2)
                    marginEnd = dp(2)
                }
            }

            if (label == "⌫") {
                // Long-press backspace = fast delete
                setupBackspaceRepeater(btn)
            } else {
                btn.setOnClickListener { onKey(label) }
            }

            row.addView(btn)
        }

        addView(row)
    }

    private var shift = false

    /**
     * Test/automation hook.
     *
     * Some devices do not expose custom keyboard children via UIAutomator dumps, and external ADB
     * tapping is flaky. Instrumentation tests should call this to drive the keyboard logic
     * deterministically.
     */
    @androidx.annotation.VisibleForTesting
    fun injectKey(label: String) {
        onKey(label)
    }

    private fun onKey(label: String) {
        val editable = target ?: return
        when (label) {
            "⌫" -> {
                val consumed = onBackspace?.invoke(currentLayout) ?: false
                if (!consumed) {
                    val len = editable.length
                    if (len > 0) editable.delete(len - 1, len)
                }
            }
            "enter" -> {
                editable.append("\n")
            }
            "space" -> {
                val consumed = onSpace?.invoke(currentLayout) ?: false
                if (!consumed) {
                    editable.append(" ")
                }
            }
            "⇧" -> {
                shift = !shift
            }
            "123" -> {
                // If we were in numeric keypad, allow switching to symbols; otherwise symbols.
                setLayout(Layout.SYMBOLS)
            }
            "abc" -> {
                // Go back based on inputMode.
                when (inputMode) {
                    InputMode.NUMBER -> setLayout(Layout.NUMERIC)
                    else -> setLayout(Layout.EN)
                }
            }
            "lang" -> {
                // In numeric/password mode, don't cycle languages.
                if (inputMode == InputMode.NUMBER) {
                    setLayout(Layout.NUMERIC)
                    return
                }
                if (inputMode == InputMode.PASSWORD) {
                    setLayout(Layout.EN)
                    return
                }

                // Cycle EN -> ZH(Pinyin) -> FR -> AR -> EN ...
                val next = when (currentLayout) {
                    Layout.EN -> Layout.ZH_PINYIN
                    Layout.ZH_PINYIN -> Layout.FR
                    Layout.FR -> Layout.AR
                    Layout.AR -> Layout.EN
                    Layout.SYMBOLS -> Layout.EN
                    Layout.NUMERIC -> Layout.EN
                }
                setLayout(next)
            }
            else -> {
                val ch = if (shift) label.uppercase() else label
                val consumed = onCommitText?.invoke(currentLayout, ch) ?: false
                if (!consumed) {
                    editable.append(ch)
                }
                // One-shot shift
                if (shift) shift = false
            }
        }
    }

    private val handler = Handler(Looper.getMainLooper())

    private fun setupBackspaceRepeater(btn: Button) {
        val initialDelayMs = 250L
        val repeatDelayMs = 50L

        var repeating = false
        val repeatRunnable = object : Runnable {
            override fun run() {
                if (!repeating) return
                onKey("⌫")
                handler.postDelayed(this, repeatDelayMs)
            }
        }

        btn.setOnTouchListener { _, event ->
            when (event.actionMasked) {
                MotionEvent.ACTION_DOWN -> {
                    // single delete immediately
                    onKey("⌫")
                    repeating = true
                    handler.postDelayed(repeatRunnable, initialDelayMs)
                    true
                }
                MotionEvent.ACTION_UP,
                MotionEvent.ACTION_CANCEL -> {
                    repeating = false
                    handler.removeCallbacks(repeatRunnable)
                    true
                }
                else -> false
            }
        }
    }

    private fun dp(v: Int): Int = (v * resources.displayMetrics.density).toInt()
}
